File: linked_lists/intersection/sol1.py
import sys

sys.path.append('../')

from sll import sll

def intersect_rec(nd1, nd2):
  if nd1.next is None and nd2.next is None:
    if nd1.data == nd2.data:
       return nd1
    else:
       return None
  
  res = intersect_rec(nd1.next, nd2.next)
  
  if res is not None:
    return nd1 if nd1.data == nd2.data else res
  return None

def intersect(sll1, sll2):
  if sll1.size < sll2.size:
    return intersect(sll2, sll1)
    
  ds = sll1.size - sll2.size
  
  nd1 = sll1.head
  for i in xrange(ds):
    nd1 = nd1.next
    
  nd2 = sll2.head
  
  return intersect_rec(nd1, nd2)
  
  
if __name__ == "__main__":
  slinked_list1 = sll.sll()
  key = 0
  value = [7, 1, 6]
  for it in value:
    slinked_list1.addLast(key, it)
    key += 1
  slinked_list2 = sll.sll()
  key = 0
  value = [5, 9, 7, 1, 6]
  for it in value:
    slinked_list2.addLast(key, it)
    key += 1
    
  print intersect(slinked_list1, slinked_list2).data

File: linked_lists/intersection/sol1.cpp
#include "sll.h"

#include <iostream>
#include <vector>

unsigned int linkedListSize(node* nd) {
  unsigned int res = 0;
  while ( nd != nullptr ) {
     ++res;
     nd = nd->next();
  }
  return res;
}

node* intersection(node* nd1, node* nd2) {
   unsigned int size1 = linkedListSize(nd1);
   unsigned int size2 = linkedListSize(nd2);
   if ( size1 < size2 ) {
     return intersection(nd2, nd1);
   }
   
   for (int i = 0; i < size1 - size2; ++i) {
      nd1 = nd1->next();
   }
   
   for ( int i = 0; i < size2; ++i ) {
      if ( nd1 == nd2 ) {
         return nd1;
      }
      nd1 = nd1->next();
      nd2 = nd2->next();
   }
   return nullptr;
}
  

int main() {
  sll slinked_list;
  std::vector<int> value{1, 2, 3, 2, 1, 1};
  int key = 0;
  for (auto& it : value) {
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  
  sll slinked_list1;
  std::vector<int> value1{1, 2, 3, 2, 1, 1};
  key = 0;
  for (auto& it : value1) {
    if (key == 0) {
      slinked_list1.insertHead(it, key);
    } else {
      slinked_list1.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  node* nd1 = slinked_list1.head();
  node* ndres = intersection(nd, nd->next()->next());

  if ( ndres == nullptr ) {
     std::cout << "no intersection";
  } else {
     std::cout << ndres->data();
  }
  
  std::cout << std::endl;
}
